qemu/tests/numa_prealloc_threads.py

```diff
@@ -43,27 +82,51 @@ def run(test, params, env):
     cpu_affinity = vm.monitor.qom_get(thread_context_device_id, "cpu-affinity")
-    affinity = str(cpu_affinity[0]) + "-" + str(cpu_affinity[-1])
+    if cpu_affinity[0] != 0:
+        affinity = str(cpu_affinity[0]) + "-" + str(cpu_affinity[-1])
+    else:
+        affinity = "0"
     test.log.debug("The affinity: %s and the expected_cpu_affinity: %s"
                    % (affinity, expected_cpu_affinity))
     if expected_cpu_affinity != affinity:
         test.fail("Test and QEMU cpu-affinity does not match!")
 
-    cmd_taskset = "taskset -c -p " + str(thread_id)
-    output = process.getoutput(cmd_taskset)
-    if not re.search(affinity, output):
-        test.fail("The affinities %s and %s do not match!"
-                  % (affinity, str(output)))
+    cmd_taskset = "taskset -c -p %s" % thread_id
+    check_affinity(affinity, cmd_taskset, "current", test)
 
     sandbox = params.get("qemu_sandbox", "on")
-    if sandbox == "on":
-        try:
-            # The command is expected to fail
-            vm.monitor.qom_set(thread_context_device_id, "cpu-affinity", cpu_affinity)
-        except QMPCmdError as e:
-            test.log.debug("The expected error message: %s and the output: %s"
-                           % (error_msg, e.data))
-            if not re.search(error_msg, str(e.data)):
-                test.fail("Can not get expected error message: %s" % error_msg)
+
+    error_context.base_context("Setting cpu-affinity: %s" % first_cpu_affinity,
+                               test.log.info)
+    try:
+        vm.monitor.qom_set(thread_context_device_id,
+                           "cpu-affinity",
+                           convert_affinity(first_cpu_affinity))
+    except QMPCmdError as e:
+        if sandbox == "off":
+            test.fail("Set cpu-affinity '%s' failed as: %s"
+                      % (first_cpu_affinity, str(e.data)))
+        if not re.search(error_msg, str(e.data)):
+            test.fail("Cannot get expected error message: %s" % error_msg)
+        test.log.debug("Get the expected error message: %s" % error_msg)
+    else:
+        if sandbox == "on":
+            test.fail("Set cpu-affinity should fail when sandbox=on")
+        affinity = first_cpu_affinity
+    check_affinity(affinity, cmd_taskset, "current", test)
+
+    if operation_type != "boot_cpu_affinity":
+        error_context.base_context("Set externally a new CPU affinity",
+                                   test.log.info)
+        cmd_taskset = "taskset -c -p %s %s" % (second_cpu_affinity,
+                                               str(thread_id))
+        error_context.context("Verify the new cpu-affinity", test.log.info)
+        check_affinity(second_cpu_affinity, cmd_taskset, "new", test)
+
+        error_context.context("Checking QEMU main thread remains untouched",
+                              test.log.info)
+        cmd_taskset = "taskset -c -p %s" % vm.get_pid()
+        check_affinity("0", cmd_taskset, "current", test)
```

Comment:
Here, `0` is the initial cpu affinity in the qemu command line, it is the `expected_cpu_affinity`  above.
Suggest use `qemu_cpu_affinity`

Reply:
Updated, thanks!

Reply:
`expected_cpu_affinity` renamed to `qemu_cpu_affinity`